src/fn/upload-custom-translations.js

```diff
@@ -81,6 +114,76 @@ function parse(filePath, options) {
   });
 }
 
+function checkTranslations(translations, lang, templatePlaceholders) {
+  let mf = null;
+  try {
+    mf = new MessageFormat(lang);
+  } catch (e) {
+    log.warn(`Cannot check '${lang}' translations: ${e.message}`);
+  }
+  const placeholders = extractPlaceholdersFromTranslations(translations);
+  let formatErrorsFound = 0,  placeholderErrorsFound = 0, emptiesFound = 0;
+  for (const [msgKey, msgSrc] of Object.entries(translations)) {
+    if (!msgSrc) {
+      emptiesFound++;
+    } else if (/{{[\s\w.#^/'|]+}}/.test(msgSrc)) {
+      if (templatePlaceholders) {
+        const placeholder = placeholders[msgKey];
+        if (placeholder) {
+          const templatePlaceholder = templatePlaceholders[msgKey];
+          if (!templatePlaceholder) {
+            log.error(`Cannot compile '${lang}' translation with key '${msgKey}' ` +
+              `has placeholders, but the key does not match any of ${EN_FILE}\n` +
+              'You can use messages-ex.properties to add placeholders missing from the reference context.');
+            placeholderErrorsFound++;
+          } else {
+            const foundAllPlaceholders = placeholder.every(el => templatePlaceholder.includes(el));
+            if (!foundAllPlaceholders) {
+              log.error(`Cannot compile '${lang}' translation with key '${msgKey}' ` +
+                `has placeholders that do not match those of ${EN_FILE}\n` +
+                'You can use messages-ex.properties to add placeholders missing from the reference context.');
+              placeholderErrorsFound++;
+            }
+          }
+        }
+      }
+    } else if (mf && typeof msgSrc === 'string') {
+      try {
+        mf.compile(msgSrc);
+      } catch (e) {
+        log.error(`Cannot compile '${lang}' translation ${msgKey} = '${msgSrc}' : ${e.message}`);
+        formatErrorsFound++;
+      }
+    }
+  }
+  if (emptiesFound > 0) {
+    log.warn(transEmptyMsg({EMPTIES: emptiesFound}) + ` '${lang}' translations`);
+  }
+  if (formatErrorsFound > 0 || placeholderErrorsFound > 0) {
+    log.error(transErrorsMsg({ERRORS: formatErrorsFound + placeholderErrorsFound})
+              + ` '${lang}' translations`);
+    process.exit(-1);
+  }
+}
+
+function extractPlaceholdersFromTranslations(translations, extraPlaceholders = {}) {
+  // Extract from github.com/medic/cht-core/blob/master/scripts/poe/lib/utils.js
+  const result = {};
+  for (const [msgKey, msgSrc] of Object.entries(translations)) {
+    let placeholders = typeof msgSrc === 'string' ? msgSrc.match(/{{[\s\w.#^/'|]+}}/g) : null;
+    if (placeholders) {
+      placeholders = placeholders
+        .sort()
+        .concat(extraPlaceholders[msgKey] ? extraPlaceholders[msgKey] : [])
+        .filter((el, i, a) => i === a.indexOf(el));
```

Comment:
If I'm reading this right this is removing duplicate placeholders. It'd be more readable to either use lodash `_.uniq` or use a Set.

Reply:
Yes, copy&paste from "poe", but due that I'll change the code there, I'll also add the changes there.

Reply:
Done.